ansible_runner/receptor_plugin.py

```diff
@@ -0,0 +1,162 @@
+                elif c_type == 'status':
+                    data = content[1]
+                    if 'status_handler' in run_options and run_options['status_handler']:
+                        run_options['status_handler'](data, None)
+                elif c_type == 'error':
+                    print("Error from remote:", content[1])
+
+    async def run_func():
+        if receptor_node_id != receptor_node:
+            add_peer_task = controller.add_peer(receptor_peer)
+            start_wait = time.time()
+            while True:
+                if add_peer_task and add_peer_task.done() and not add_peer_task.result():
+                    return False
+                if controller.receptor.router.node_is_known(receptor_node):
+                    break
+                if time.time() - start_wait > 5:
+                    if not add_peer_task.done():
+                        add_peer_task.cancel()
+                    return False
+                await asyncio.sleep(0.1)
+        await controller.send(payload=tmpf.name, recipient=receptor_node, directive='ansible_runner:execute')
+        await controller.loop.create_task(read_responses())
+
+    remote_options = {key: value for key, value in run_options.items() if key in remote_run_options}
+
+    with tempfile.NamedTemporaryFile(suffix='.tgz') as tmpf:
+
+        # Create tar file
+        with tarfile.open(fileobj=tmpf, mode='w:gz') as tar:
+            if 'private_data_dir' in run_options:
+                tar.add(run_options['private_data_dir'], arcname='')
+            kwargs = json.dumps(remote_options, cls=UUIDEncoder)
+            ti = tarfile.TarInfo('kwargs')
+            ti.size = len(kwargs)
+            ti.mtime = time.time()
+            tar.addfile(ti, io.BytesIO(kwargs.encode('utf-8')))
+        tmpf.flush()
+
+        # Run the job via Receptor
+        if receptor_node_id:
+            receptor_args = f"--node-id {receptor_node_id} node --server-disable".split()
+        else:
+            receptor_args = "node --server-disable".split()
+        config = ReceptorConfig(receptor_args)
+        config._is_ephemeral = True
+        controller = Controller(config)
+        try:
+            controller.run(run_func)
+        finally:
+            controller.cleanup_tmpdir()
+
+        return type('Receptor_Runner_Result', (), {'rc': 0})
+
+
+# We set these parameters locally rather than using receptor.plugin_utils
+# because this still needs to parse even when our import of receptor failed.
+def receptor_plugin_export(func):
+    if receptor_import:
+        func.receptor_export = True
+        func.payload_type = receptor.BUFFER_PAYLOAD
+    return func
+
+
+@receptor_plugin_export
+def execute(message, config, result_queue):
+    private_dir = None
+    try:
+        private_dir = tempfile.TemporaryDirectory()
+        with tarfile.open(fileobj=message.fp, mode='r:gz') as tar:
+            tar.extractall(path=private_dir.name)
```

Comment:
This is dangerous if you don't completely trust the creator of the tar file. 
```
2. Symlink file attack:

    foo links to /etc/passwd.
    Another member named foo follows, its data overwrites the target file's data.
```
https://bugs.python.org/issue21109
https://bugs.python.org/issue21109#msg215656

Reply:
Worked around this by changing from `tarfile` to `zipfile`.  Oddly, the Python `zipfile` class has built-in protection against these traversal attacks, even though `tarfile` does not.